## Импорт пандаса

In [59]:
import pandas as pd

## Считывание файлов

In [2]:
assessments = pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-d-pak/first_pr_files/assessments.csv')

In [3]:
courses = pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-d-pak/first_pr_files/courses.csv')

In [4]:
student_assessment = pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-d-pak/first_pr_files/studentAssessment.csv')

In [5]:
student_registration = pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-d-pak/first_pr_files/studentRegistration.csv')

In [6]:
assessments.head()

code_module code_presentation  id_assessment assessment_type   date  weight
0         AAA             2013J           1752             TMA   19.0    10.0
1         AAA             2013J           1753             TMA   54.0    20.0
2         AAA             2013J           1754             TMA  117.0    20.0
3         AAA             2013J           1755             TMA  166.0    20.0
4         AAA             2013J           1756             TMA  215.0    30.0

In [7]:
student_assessment.head()

id_assessment  id_student  date_submitted  is_banked  score
0           1752       11391              18          0   78.0
1           1752       28400              22          0   70.0
2           1752       31604              17          0   72.0
3           1752       32885              26          0   69.0
4           1752       38053              19          0   79.0

In [8]:
courses.head()

code_module code_presentation  module_presentation_length
0         AAA             2013J                         268
1         AAA             2014J                         269
2         BBB             2013J                         268
3         BBB             2014J                         262
4         BBB             2013B                         240

In [9]:
student_registration.head()

code_module code_presentation  id_student  date_registration  \
0         AAA             2013J       11391             -159.0   
1         AAA             2013J       28400              -53.0   
2         AAA             2013J       30268              -92.0   
3         AAA             2013J       31604              -52.0   
4         AAA             2013J       32885             -176.0   

   date_unregistration  
0                  NaN  
1                  NaN  
2                 12.0  
3                  NaN  
4                  NaN

## Что считать курсом?

In [10]:
student_registration.code_presentation.unique()

array(['2013J', '2014J', '2013B', '2014B'], dtype=object)

Взаимосвязи между кодом предмета и семестром не наблюдается:

In [11]:
student_registration \
.groupby(['code_module', 'code_presentation'], as_index = False) \
.agg({'date_registration': 'count'})

code_module code_presentation  date_registration
0          AAA             2013J                383
1          AAA             2014J                365
2          BBB             2013B               1766
3          BBB             2013J               2235
4          BBB             2014B               1608
5          BBB             2014J               2291
6          CCC             2014B               1931
7          CCC             2014J               2495
8          DDD             2013B               1299
9          DDD             2013J               1933
10         DDD             2014B               1225
11         DDD             2014J               1800
12         EEE             2013J               1050
13         EEE             2014B                694
14         EEE             2014J               1188
15         FFF             2013B               1611
16         FFF             2013J               2282
17         FFF             2014B               1495
18         FFF             2014J               2363
19         GGG             2013J                952
20         GGG             2014B                833
21         GGG             2014J                749

Экзамены были по каждому предмету и в каждом семестре:

In [12]:
assessments \
.groupby(['code_module', 'code_presentation'], as_index = False) \
.agg({'id_assessment': 'count'})

code_module code_presentation  id_assessment
0          AAA             2013J              6
1          AAA             2014J              6
2          BBB             2013B             12
3          BBB             2013J             12
4          BBB             2014B             12
5          BBB             2014J              6
6          CCC             2014B             10
7          CCC             2014J             10
8          DDD             2013B             14
9          DDD             2013J              7
10         DDD             2014B              7
11         DDD             2014J              7
12         EEE             2013J              5
13         EEE             2014B              5
14         EEE             2014J              5
15         FFF             2013B             13
16         FFF             2013J             13
17         FFF             2014B             13
18         FFF             2014J             13
19         GGG             2013J             10
20         GGG             2014B             10
21         GGG             2014J             10

Т.к.code_module - код предмета, по которому в каждом семестре (если была регистрация) проводится экзамен, курсом я буду считать code_module

## Сколько студентов успешно сдали только один курс?

Объединяю таблицы, чтобы отфильтровать строки только с экзаменами:

In [13]:
full_assessments = student_assessment.merge(assessments)
full_assessments

id_assessment  id_student  date_submitted  is_banked  score  \
0                1752       11391              18          0   78.0   
1                1752       28400              22          0   70.0   
2                1752       31604              17          0   72.0   
3                1752       32885              26          0   69.0   
4                1752       38053              19          0   79.0   
...               ...         ...             ...        ...    ...   
173907          37443      527538             227          0   60.0   
173908          37443      534672             229          0  100.0   
173909          37443      546286             215          0   80.0   
173910          37443      546724             230          0  100.0   
173911          37443      558486             224          0   80.0   

       code_module code_presentation assessment_type   date  weight  
0              AAA             2013J             TMA   19.0    10.0  
1              AAA             2013J             TMA   19.0    10.0  
2              AAA             2013J             TMA   19.0    10.0  
3              AAA             2013J             TMA   19.0    10.0  
4              AAA             2013J             TMA   19.0    10.0  
...            ...               ...             ...    ...     ...  
173907         GGG             2014J             CMA  229.0     0.0  
173908         GGG             2014J             CMA  229.0     0.0  
173909         GGG             2014J             CMA  229.0     0.0  
173910         GGG             2014J             CMA  229.0     0.0  
173911         GGG             2014J             CMA  229.0     0.0  

[173912 rows x 10 columns]

На один курс могло приходиться больше одного экзамена, поэтому считать по кол-ву сданных экзаменов - не лучшая идея:

In [14]:
full_assessments \
.query('assessment_type == "Exam"') \
.groupby('code_module', as_index = False) \
.agg({'id_assessment' : 'nunique'})

code_module  id_assessment
0         CCC              2
1         DDD              4

Создаю таблицу с данными только об экзаменах:

In [15]:
exams = full_assessments.query('assessment_type == "Exam"')

Добавляю столбец, который отражает, сдал ли студент экзамен:

In [16]:
exams = exams.assign(success = exams.score >= 40)
exams.head()

id_assessment  id_student  date_submitted  is_banked  score  \
52923          24290      558914             230          0   32.0   
52924          24290      559706             234          0   78.0   
52925          24290      559770             230          0   54.0   
52926          24290      560114             230          0   64.0   
52927          24290      560311             234          0  100.0   

      code_module code_presentation assessment_type  date  weight  success  
52923         CCC             2014B            Exam   NaN   100.0    False  
52924         CCC             2014B            Exam   NaN   100.0     True  
52925         CCC             2014B            Exam   NaN   100.0     True  
52926         CCC             2014B            Exam   NaN   100.0     True  
52927         CCC             2014B            Exam   NaN   100.0     True

Сколько студентов успешно завершили только один курс:

In [17]:
exams \
.query('success == True') \
.groupby('id_student', as_index = False) \
.agg({'code_module':'count'}) \
.query('code_module == 1') \
.shape[0]

3802

Ответ по заданию 1: **3802** студента сдали только один курс

## Выяви самый сложный и самый простой экзамен: найди курсы и экзамены в рамках курса, которые обладают самой низкой и самой высокой завершенностью

***Завершенность = количество успешных экзаменов / количество всех попыток сдать экзамен***

Количество успешных попыток в разбивке по курсу и экзамену:

In [18]:
success = exams \
.query('success == True') \
.groupby(['code_module', 'id_assessment'], as_index = False) \
.agg({'id_student' : 'count'}) \
.rename(columns={'id_student' : 'success_count'})
success

code_module  id_assessment  success_count
0         CCC          24290            664
1         CCC          24299           1019
2         DDD          25340            504
3         DDD          25354            878
4         DDD          25361            485
5         DDD          25368            842

Количество всех попыток:

In [19]:
total_attempts = exams \
.groupby(['code_module', 'id_assessment'], as_index = False) \
.agg({'id_student' : 'count'}) \
.rename(columns={'id_student' : 'total_attempts'})
total_attempts

code_module  id_assessment  total_attempts
0         CCC          24290             747
1         CCC          24299            1168
2         DDD          25340             602
3         DDD          25354             968
4         DDD          25361             524
5         DDD          25368             950

Объединяю таблицы:

In [20]:
df_exams = success.merge(total_attempts, on=['code_module', 'id_assessment'])
df_exams

code_module  id_assessment  success_count  total_attempts
0         CCC          24290            664             747
1         CCC          24299           1019            1168
2         DDD          25340            504             602
3         DDD          25354            878             968
4         DDD          25361            485             524
5         DDD          25368            842             950

Создаю финальную таблицу с новым столбцом, содержащим процент завершенности:

In [21]:
complete_ex_df = df_exams.assign(complete = round(df_exams.success_count / df_exams.total_attempts * 100, 2)) \
.groupby(['code_module', 'id_assessment'], as_index = False) \
.agg({'complete' : 'mean'})
complete_ex_df

code_module  id_assessment  complete
0         CCC          24290     88.89
1         CCC          24299     87.24
2         DDD          25340     83.72
3         DDD          25354     90.70
4         DDD          25361     92.56
5         DDD          25368     88.63

Самый сложный экзамен:

In [22]:
complete_ex_df[complete_ex_df['complete'] == complete_ex_df['complete'].min()]

code_module  id_assessment  complete
2         DDD          25340     83.72

Самый легкий экзамен:

In [23]:
complete_ex_df[complete_ex_df['complete'] == complete_ex_df['complete'].max()]

code_module  id_assessment  complete
4         DDD          25361     92.56

Завершаемость каждого курса в целом:

In [24]:
complete_ex_df \
.groupby('code_module', as_index = False) \
.agg({'complete': 'mean'})

code_module  complete
0         CCC   88.0650
1         DDD   88.9025

Ответы по заданию 2:  
Самый сложный экзамен: **25340**  
Самый простой экзамен: **25361**  
Курс с наиболее высокой завершенностью: **DDD**  
Курс с наименее высокой завершенностью: **CCC**  

## По каждому предмету определи средний срок сдачи экзаменов (под сдачей понимаем последнее успешное прохождение экзамена студентом)

Создаю таблицу с последней успешной сдачей экзамена студентом (т.е. максимальной):

In [25]:
last_exams = exams \
.query('success == True') \
.groupby(['code_module', 'id_student'], as_index = False) \
.agg({'date_submitted' : 'max'})
last_exams

code_module  id_student  date_submitted
0            CCC       23698             243
1            CCC       27116             243
2            CCC       28787             243
3            CCC       29420             244
4            CCC       29639             248
...          ...         ...             ...
4384         DDD     2693243             230
4385         DDD     2694886             236
4386         DDD     2694933             230
4387         DDD     2695608             237
4388         DDD     2698251             234

[4389 rows x 3 columns]

Дополняю ее общей информацией по экзаменам, чтобы фильтровать данные:

In [26]:
full_last_exams = last_exams.merge(exams)
full_last_exams

code_module  id_student  date_submitted  id_assessment  is_banked  score  \
0            CCC       23698             243          24299          0   80.0   
1            CCC       27116             243          24299          0   96.0   
2            CCC       28787             243          24299          0   44.0   
3            CCC       29420             244          24299          0   62.0   
4            CCC       29639             248          24299          0   74.0   
...          ...         ...             ...            ...        ...    ...   
4384         DDD     2693243             230          25340          0   82.0   
4385         DDD     2694886             236          25361          0   69.0   
4386         DDD     2694933             230          25340          0   73.0   
4387         DDD     2695608             237          25354          0   73.0   
4388         DDD     2698251             234          25361          0   44.0   

     code_presentation assessment_type   date  weight  success  
0                2014J            Exam    NaN   100.0     True  
1                2014J            Exam    NaN   100.0     True  
2                2014J            Exam    NaN   100.0     True  
3                2014J            Exam    NaN   100.0     True  
4                2014J            Exam    NaN   100.0     True  
...                ...             ...    ...     ...      ...  
4384             2013B            Exam  240.0   100.0     True  
4385             2014B            Exam  241.0   100.0     True  
4386             2013B            Exam  240.0   100.0     True  
4387             2013J            Exam  261.0   100.0     True  
4388             2014B            Exam  241.0   100.0     True  

[4389 rows x 11 columns]

Считаю среднее количество дней, требующихся для сдачи экзамена:

In [27]:
full_last_exams \
.query('success == True') \
.groupby('code_module', as_index = False) \
.agg({'date_submitted' : 'mean'}) 

code_module  date_submitted
0         CCC      239.345217
1         DDD      237.978566

Ответ по заданию 3:  
Средний срок сдачи экзаменов для курса CCC: **239 дней**  
Средний срок сдачи экзаменов для курса DDD: **238 дней**

## Выяви самые популярные предметы (ТОП-3) по количеству регистраций на них, а также предметы с самым большим оттоком (ТОП-3)

Считаю количество регистраций на каждый курс(предмет), сортирую и выбираю ТОП-3:

In [28]:
student_registration \
.value_counts('code_module') \
.head(3) \
.to_frame(name='registrations') \
.reset_index() 

code_module  registrations
0         BBB           7909
1         FFF           7762
2         DDD           6272

***Отток = количество студентов, отменивших регистрацию / количество студентов, которые регистрировались***

Сколько всего людей зарегестрировалось на какой-либо курс:

In [29]:
total = student_registration \
.groupby('code_module', as_index = False) \
.agg({'date_registration' : 'count'}) \
.rename(columns = {'date_registration': 'total'})
total

code_module  total
0         AAA    748
1         BBB   7900
2         CCC   4426
3         DDD   6257
4         EEE   2932
5         FFF   7751
6         GGG   2534

Сколько отменили регистрацию (не берем в расчет значения NaN, т.к. эти студенты окончили курс):

In [30]:
unreg = student_registration \
.query('date_unregistration != "NaN"') \
.groupby('code_module', as_index = False) \
.agg({'date_unregistration' : 'count'}) \
.rename(columns = {'date_unregistration': 'unreg'})
unreg

code_module  unreg
0         AAA    126
1         BBB   2377
2         CCC   1947
3         DDD   2235
4         EEE    718
5         FFF   2380
6         GGG    289

Объединяю таблицы:

In [31]:
churn_df = total.merge(unreg)

Создаю новый столбец с показателем оттока студентов, сортирую по самому высокому оттоку:

In [32]:
churn_df \
.assign(churn = round(churn_df.unreg / churn_df.total * 100, 2)) \
.sort_values('churn', ascending = False) \
.reset_index(drop=True) \
.head(3)

code_module  total  unreg  churn
0         CCC   4426   1947  43.99
1         DDD   6257   2235  35.72
2         FFF   7751   2380  30.71

Ответы по заданию 4:  
ТОП-3 предметов по количеству регистраций: **BBB, FFF, DDD**  
ТОП-3 предметов с самым большим оттоком: **CCC, DDD, FFF**  

## Используя pandas, в период с начала 2013 по конец 2014 выяви семестр с самой низкой завершенностью курсов и самыми долгими средними сроками сдачи курсов

In [33]:
exams.code_presentation.unique()

array(['2014B', '2014J', '2013B', '2013J'], dtype=object)

Данные есть только за  2013 и 2014 год, фильтрация не требуется.

Количество успешных попыток в разбивке по семестрам:

In [34]:
success_sem = exams \
.query('success == True') \
.groupby('code_presentation', as_index = False) \
.agg({'id_student' : 'count'}) \
.rename(columns={'id_student' : 'success_count'})
success_sem

code_presentation  success_count
0             2013B            504
1             2013J            878
2             2014B           1149
3             2014J           1861

Всего попыток:

In [35]:
total_attempts_sem = exams \
.groupby('code_presentation', as_index = False) \
.agg({'id_student' : 'count'}) \
.rename(columns={'id_student' : 'total_attempts'})
total_attempts_sem

code_presentation  total_attempts
0             2013B             602
1             2013J             968
2             2014B            1271
3             2014J            2118

Объединяю таблицы:

In [36]:
df_semester = success_sem.merge(total_attempts_sem, on='code_presentation')

Добавляю новый столбец с показателем завершаемости:

In [37]:
df_semester = df_semester \
.assign(complete = round(df_semester.success_count / df_semester.total_attempts * 100, 2)) 

Семестр с самой низкой завершаемостью:

In [38]:
df_semester[df_semester['complete'] == df_semester.complete.min()]

code_presentation  success_count  total_attempts  complete
0             2013B            504             602     83.72

Средние сроки сдачи экзаменов по курсам и семестрам:

In [39]:
avg_date = full_last_exams \
.groupby(['code_presentation', 'code_module'], as_index = False) \
.agg({'date_submitted' : 'mean'}) \
.rename(columns={'date_submitted' : 'avg_date'})
avg_date

code_presentation code_module    avg_date
0             2013B         DDD  230.166998
1             2013J         DDD  239.510274
2             2014B         CCC  231.581325
3             2014B         DDD  234.936082
4             2014J         CCC  244.404318
5             2014J         DDD  242.804038

Самые долгие средние сроки сдачи курсов:

In [40]:
avg_date[avg_date['avg_date'] == avg_date.avg_date.max()]

code_presentation code_module    avg_date
4             2014J         CCC  244.404318

Ответы по заданию 5:  
Семестр с самой низкой завершенностью курсов: **2013B**  
Семестр с самыми долгими средними сроками сдачи курсов: **2014J**  

## Используя python, построй адаптированные RFM-кластеры студентов, чтобы качественно оценить свою аудиторию

Среднее время сдачи одного экзамена:

In [41]:
r = exams \
.groupby(['id_student'], as_index = False) \
.agg({'date_submitted' : 'mean'}) \
.rename(columns = {'date_submitted' : 'Recency'})
r

id_student  Recency
0          23698    243.0
1          24213    236.0
2          27116    243.0
3          28046    237.0
4          28787    243.0
...          ...      ...
4628     2694886    236.0
4629     2694933    230.0
4630     2695608    237.0
4631     2697181    230.0
4632     2698251    234.0

[4633 rows x 2 columns]

Количество успешных попыток сдать экзамен для каждого студента:

In [42]:
student_success = full_last_exams \
.groupby(['id_student'], as_index = False) \
.agg({'id_assessment' : 'count'}) \
.rename(columns={'id_assessment' : 'success_count'})
student_success

id_student  success_count
0          23698              1
1          24213              1
2          27116              1
3          28046              1
4          28787              1
...          ...            ...
4092     2694886              1
4093     2694933              1
4094     2695608              1
4095     2697181              1
4096     2698251              1

[4097 rows x 2 columns]

Всего попыток для каждого студента:

In [43]:
student_total_attempts = exams \
.groupby(['id_student'], as_index = False) \
.agg({'id_assessment' : 'count'}) \
.rename(columns={'id_assessment' : 'total_attempts'})
student_total_attempts

id_student  total_attempts
0          23698               1
1          24213               1
2          27116               1
3          28046               1
4          28787               1
...          ...             ...
4628     2694886               1
4629     2694933               1
4630     2695608               1
4631     2697181               1
4632     2698251               1

[4633 rows x 2 columns]

Объединяю таблицы:

In [44]:
f = student_success.merge(student_total_attempts, how='outer', on='id_student')
f = f.fillna(0)
f

id_student  success_count  total_attempts
0          23698            1.0               1
1          24213            1.0               1
2          27116            1.0               1
3          28046            1.0               1
4          28787            1.0               1
...          ...            ...             ...
4628     2611779            0.0               1
4629     2615956            0.0               1
4630     2655683            0.0               1
4631     2661870            0.0               1
4632     2691861            0.0               1

[4633 rows x 3 columns]

Рассчитываю показатель завершенности курсов каждым студентом:

In [45]:
f = f.assign(Frequency = round(f.success_count / f.total_attempts * 100, 2)) \
.drop(['success_count', 'total_attempts'], axis=1)
f

id_student  Frequency
0          23698      100.0
1          24213      100.0
2          27116      100.0
3          28046      100.0
4          28787      100.0
...          ...        ...
4628     2611779        0.0
4629     2615956        0.0
4630     2655683        0.0
4631     2661870        0.0
4632     2691861        0.0

[4633 rows x 2 columns]

Среднее количество баллов, получаемое за экзамен:

In [46]:
m = exams \
.groupby('id_student', as_index = False) \
.agg({'score' : 'mean'}) \
.rename(columns = {'score' : 'Monetary'})
m

id_student  Monetary
0          23698      80.0
1          24213      58.0
2          27116      96.0
3          28046      40.0
4          28787      44.0
...          ...       ...
4628     2694886      69.0
4629     2694933      73.0
4630     2695608      73.0
4631     2697181      80.0
4632     2698251      44.0

[4633 rows x 2 columns]

Объединяю все таблицы для расчета метрик:

In [47]:
rf = r.merge(f)
rfm = rf.merge(m)
rfm

id_student  Recency  Frequency  Monetary
0          23698    243.0      100.0      80.0
1          24213    236.0      100.0      58.0
2          27116    243.0      100.0      96.0
3          28046    237.0      100.0      40.0
4          28787    243.0      100.0      44.0
...          ...      ...        ...       ...
4628     2694886    236.0      100.0      69.0
4629     2694933    230.0      100.0      73.0
4630     2695608    237.0      100.0      73.0
4631     2697181    230.0      100.0      80.0
4632     2698251    234.0      100.0      44.0

[4633 rows x 4 columns]

Границы для интерпретации кластеров:

In [48]:
quintiles = rfm[['Recency', 'Monetary']].quantile([.2, .4, .6, .8]).to_dict()
quintiles

{'Recency': {0.2: 231.0, 0.4: 237.0, 0.6: 242.0, 0.8: 243.5},
 'Monetary': {0.2: 46.200000000000045, 0.4: 60.0, 0.6: 72.0, 0.8: 84.0}}

In [58]:
rfm.Frequency.unique()

array([100.,   0.,  50.])

Есть три вида завершенности курсов (сдал с первого раза, со второго, не сдал).  
Метрика, подобранная для среднего времени сдачи и среднего балла, не подойдет для завершенности.

Метрика для оценки среднего срока сдачи экзамена: чем быстрее студент сдал экзамен, тем лучше, и тем выше оценка:

In [50]:
def r_score(x):
    if x <= quintiles['Recency'][.2]:
        return 5
    elif x <= quintiles['Recency'][.4]:
        return 4
    elif x <= quintiles['Recency'][.6]:
        return 3
    elif x <= quintiles['Recency'][.8]:
        return 2
    else:
        return 1

Метрика для оценки завершенности студентами курсов: если сдал с первого раза, оценка выше:

In [51]:
def f_score(x):
    if x == 100:
        return 2
    if x == 50:
        return 1
    else:
        return 0

Метрика для оценки среднего балла, получаемого за экзамен: чем выше средний балл, тем выше оценка:

In [52]:
def m_score(x):
    if x <= quintiles['Monetary'][.2]:
        return 1
    elif x <= quintiles['Monetary'][.4]:
        return 2
    elif x <= quintiles['Monetary'][.6]:
        return 3
    elif x <= quintiles['Monetary'][.8]:
        return 4
    else:
        return 5

Применяю функции к таблице, рассчитывая метрики для каждого студента:

In [53]:
rfm['R'] = rfm['Recency'].apply(lambda x: r_score(x))
rfm['F'] = rfm['Frequency'].apply(lambda x: f_score(x))
rfm['M'] = rfm['Monetary'].apply(lambda x: m_score(x))
rfm

id_student  Recency  Frequency  Monetary  R  F  M
0          23698    243.0      100.0      80.0  2  2  4
1          24213    236.0      100.0      58.0  4  2  2
2          27116    243.0      100.0      96.0  2  2  5
3          28046    237.0      100.0      40.0  4  2  1
4          28787    243.0      100.0      44.0  2  2  1
...          ...      ...        ...       ... .. .. ..
4628     2694886    236.0      100.0      69.0  4  2  3
4629     2694933    230.0      100.0      73.0  5  2  4
4630     2695608    237.0      100.0      73.0  4  2  4
4631     2697181    230.0      100.0      80.0  5  2  4
4632     2698251    234.0      100.0      44.0  4  2  1

[4633 rows x 7 columns]

Кластеры построены, оцениваю самые частые значения каждого из них:

In [54]:
r_analysis = rfm \
.groupby(['R'], as_index = False) \
.agg({'id_student': 'count'}) \
.rename(columns={'id_student': 'cnt'}) \
.sort_values('cnt', ascending = False)
r_analysis.assign(percent = round(r_analysis.cnt / 4633 * 100, 2))

R   cnt  percent
3  4  1144    24.69
4  5   957    20.66
2  3   917    19.79
0  1   916    19.77
1  2   699    15.09

In [55]:
f_analysis = rfm \
.groupby(['F'], as_index = False) \
.agg({'id_student': 'count'}) \
.rename(columns={'id_student': 'cnt'}) \
.sort_values('cnt', ascending = False)
f_analysis.assign(percent = round(f_analysis.cnt / 4633 * 100, 2))

F   cnt  percent
2  2  4066    87.76
0  0   536    11.57
1  1    31     0.67

In [56]:
m_analysis = rfm \
.groupby(['M'], as_index = False) \
.agg({'id_student': 'count'}) \
.rename(columns={'id_student': 'cnt'}) \
.sort_values('cnt', ascending = False)
m_analysis.assign(percent = round(m_analysis.cnt / 4633 * 100, 2))

M   cnt  percent
1  2  1043    22.51
0  1   927    20.01
4  5   911    19.66
3  4   895    19.32
2  3   857    18.50